### PR TITLE
Utility to check that same version of all modules is used

### PR DESCRIPTION
**Pull request by patriknw**
* Stolen from Cinnamon
* Can be used from outside of Akka, e.g. Akka HTTP or Lagom

Can't write an integration test here since it only finds things in real jars. I tried it out in an sample, which turned out to have mixed versions :) `akka-sample-persistence-dc-scala`

Perhaps a multi-jvm test would work?

cc @bantonsson

### CODE REVIEW

File: akka-actor/src/main/scala/akka/util/ManifestInfo.scala [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import akka.actor.Extension
+import akka.actor.ExtensionId
+import akka.actor.ExtensionIdProvider
+import akka.annotation.InternalApi
+import akka.event.Logging
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object ManifestInfo extends ExtensionId[ManifestInfo] with ExtensionIdProvider {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ManifestInfo.scala]: make public

File: akka-actor/src/main/scala/akka/util/ManifestInfo.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    override def toString: String = version
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class ManifestInfo(val system: ExtendedActorSystem) extends Extension {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ManifestInfo.scala]: public

File: akka-actor/src/main/scala/akka/util/ManifestInfo.scala [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        }
+      }
+    } catch {
+      case ioe: IOException ⇒
+        Logging(system, getClass).warning("Could not read manifest information. {}", ioe)
+    }
+    manifests
+  }
+
+  def checkSameVersion(dependencies: Set[String]): Unit = {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ManifestInfo.scala]: make log message more customizable, e.g. for usage from Play

**Comment by patriknw** [comment]
completed this @marcospereira  @renatocaval

**Review by marcospereira — APPROVED** [review]
LGTM. :-)

**Review by chbatey — APPROVED** [review]
LGTM

**Review by ktoso — APPROVED** [review]
Lookin good!

File: akka-actor/src/main/scala/akka/util/ManifestInfo.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    override def toString: String = version
+  }
+}
+
+/**
+ * Utility that extracts [[ManifestInfo#Version]] information from META-INF/MANIFEST.MF in jar files on the classpath.
+ * Note that versions can only be found in ordinary jar files, for example not in "fat jars' assembled from
+ * many jar files.
+ */
+class ManifestInfo(val system: ExtendedActorSystem) extends Extension {
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ManifestInfo.scala]: final?

File: akka-actor/src/main/scala/akka/util/ManifestInfo.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  override def get(system: ActorSystem): ManifestInfo = super.get(system)
+
+  override def lookup(): ManifestInfo.type = ManifestInfo
+
+  override def createExtension(system: ExtendedActorSystem): ManifestInfo = new ManifestInfo(system)
+
+  /**
+   * Comparable version information
+   */
+  class Version(val version: String) extends Comparable[Version] {
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor/src/main/scala/akka/util/ManifestInfo.scala]: yes, I'll fix

**Comment by raboof** [comment]
Keep in mind that this warning will not show up when using a 'fat jar': in that case the `META-INF/MANIFEST.MF` is lost...

**Comment by patriknw** [comment]
I know, another reason I don’t like such jars. Hopefully it’s only used for deployments and the warning is noticed before going to production.

**Pull request merged**